iroh-dns blog post (#122)
n0-computer/iroh.computer: src/app/blog/iroh-dns/page.mdx

import Image from 'next/image'

import { BlogPostLayout } from '@/components/BlogPostLayout'
import { MotionCanvas } from '@/components/MotionCanvas'

export const post = {
  draft: false,

>   author: 'ramfox',

Reviewer: ```suggestion
  author: 'ramfox', 'frando',
```

  date: '2024-05-06',
  title: 'Dial by NodeID, no address required!',
  description: 'iroh-dns uses domain name servers to resolve dialing details, so you can dial any device by it\'s node identifier',
}

export const metadata = {
  title: post.title,
  description: post.description,
}

export default (props) => <BlogPostLayout article={post} {...props} />

## The motivation

> We want iroh to establish connections with as little friction as possible. We’ve been chipping away at this goal since our inception. Our first big push toward this goal was to add holepunching into iroh, removing the need for devs to worry about opening up ports on their servers or resigning themselves to only making peer-to-peer connections to computers inside their local NAT.

Reviewer: ```suggestion
We want iroh to establish connections with as little friction as possible. Our first big push toward chipping away at this goal was adding holepunching into iroh. Now, devs no longer need to worry about opening up ports on their servers / firewalls or be resigned to only creating peer-to-peer connections to computers inside their local NAT.
```

> But even with holepunching, you need to know *where* to dial. Dialing a node in Iroh needs either an IP address to talk to, or the URL of a relay to which the remote node is connected. To make things easier, we added [tickets](/docs/concepts/tickets) early-on. Tickets are easily encodable bytestrings that contain a node id, socket addresses and a relay URL. Tickets work well, but they are long - and they expire: People change networks or relays, and then old tickets cannot find you anymore. So while tickets work well in some scenarios, they are not frictionless or universally usable. 

Reviewer: ```suggestion
But even with holepunching, you need to know *where* to dial. Dialing a node in iroh needs either an IP address to talk to, or the URL of a relay to which the remote node is connected. To make things easier, we added [tickets](/docs/concepts/tickets) early-on. Tickets are easily encodable bytestrings that contain a node id, socket addresses and a relay URL. Tickets work well, but they are long - and they expire: People change networks or relays, and then old tickets cannot find you anymore. So while tickets work well in some scenarios, they are not frictionless or universally usable. 
```

## The approach

We are using two standards: the domain name system [(DNS)](https://datatracker.ietf.org/doc/html/rfc1035) and [Pkarr](https://app.pkarr.org/). DNS is typically used for resolving domain names into their associated IP addresses. But DNS actually relates _domain names_ to [_resource records_](https://datatracker.ietf.org/doc/html/rfc1035#section-3.2), of which IP addresses are one type. Pkarr allows us to publish DNS records that resolve **elliptic curve keys** (not domain names) to resource records instead. Our **Node IDs** in iroh are elliptic curve public keys. This means we can publish an association between iroh Node IDs and some resource records. Importantly, these records are signed, so that you can verify that the record was actually published by the node with the given Node ID.

> As long as iroh has a Node ID and it’s associated relay URL (the address of the relay server that node uses to hole-punch and proxy relay packets), we can dial that node. So the Pkarr packet currently only needs to contain the Node ID and the relay URL of it’s preferred relay server (which we call its “home relay”). When Pkarr publishing is enabled on your iroh node, your node will create a Pkarr packet with its Node ID and relay URL, sign it, and defaults to publishing on an  `iroh-dns` server that is run by number0.

Reviewer: ```suggestion
As long as iroh has a Node ID and its associated relay URL (the address of the relay server that node uses to hole-punch and proxy relay packets), we can dial that node. So the Pkarr packet currently only needs to contain the Node ID and the relay URL of its preferred relay server (which we call its “home relay”). When Pkarr publishing is enabled on your iroh node, your node will create a Pkarr packet with its Node ID and relay URL, sign it, and defaults to publishing on an  `iroh-dns` server that is run by number0.
```

  <MotionCanvas src='/motion-canvas/Publishing.project.js' />
</div>

Nodes announce their address information in [Pkarr](https://github.com/Nuhvi/pkarr/) signed packets. The TXT records, as described below, are added to the `answers` section of a DNS server. Their name must be `_iroh.<node-id>.` The encoded node ID must be the root name, no other origin but `.` (the single dot) is permitted.

Those packets are published to a Pkarr relay server, which is a HTTP service handling PUT requests with the signed packets. Iroh’s Pkarr server is [`iroh-dns-server`](https://crates.io/crates/iroh-dns-server), which serves the received records over DNS. Pkarr packets can also be published onto the bittorrent mainline DHT, as specified by Pkarr (this is not yet supported in Iroh natively, see below for our plans).

DNS servers that support this spec will receive these Pkarr signed packets, check their signature and format validity, and then serve each contained record, with the DNS server's configured *origin domain* appended to all record names.

## Node discovery via DNS

<div className='not-prose mb-5'>
  <MotionCanvas src='/motion-canvas/Discovery.project.js' />
</div>

> When connecting to an unknown `NodeId`, the DNS discovery mechanism in Iroh will perform a DNS query to discover relays or addresses for the node. The DNS discovery is configured with a *origin domain* (which defaults to *dns.iroh.link*, a server run by n0). Iroh will then perform a DNS query through the configured DNS resolver (which defaults to using the host system's nameservers):

Reviewer: ```suggestion
When connecting to an unknown `NodeId`, the DNS discovery mechanism in iroh will perform a DNS query to discover relays or addresses for the node. The DNS discovery is configured with a *origin domain* (which defaults to *dns.iroh.link*, a server run by n0). Iroh will then perform a DNS query through the configured DNS resolver (which defaults to using the host system's nameservers):
```
